README.md

# send-stream

[![NPM](https://img.shields.io/npm/v/send-stream.svg)](https://www.npmjs.com/package/send-stream)
[![Build Status](https://github.com/nicolashenry/send-stream/workflows/CI/badge.svg?branch=master)](https://github.com/nicolashenry/send-stream/actions?query=workflow%3ACI+branch%3Amaster)
[![Coverage Status](https://coveralls.io/repos/github/nicolashenry/send-stream/badge.svg?branch=master)](https://coveralls.io/github/nicolashenry/send-stream?branch=master)

`send-stream` is a library for streaming files from the file system or any other source.

It supports partial responses (Ranges including multipart), conditional-GET negotiation (If-Match, If-Unmodified-Since, If-None-Match, If-Modified-Since) and precompressed content
encoding negociation.

It also have high test coverage, typescript typings
and has [multiple examples](examples/) using Express, Koa, Fastify or pure NodeJS http/http2 modules.

## Installation

This is a [Node.js](https://nodejs.org/en/) module available through the
[npm registry](https://www.npmjs.com/). Installation is done using the
[`npm install` command](https://docs.npmjs.com/getting-started/installing-npm-packages-locally):

```bash
$ npm install send-stream
```

## Getting start

Serve all files from a directory (also serve index.html from directories on trailing slash) with Fastify, Koa or Express

See [examples](#examples) for more advanced usages.

Using Fastify (v3.x.x):
```js
const path = require('path');
const fastify = require('fastify');
const { FileSystemStorage } = require('send-stream');

const app = fastify();
const storage = new FileSystemStorage(path.join(__dirname, 'assets'), { onDirectory: 'serve-index' });

app.route({
  method: ['HEAD', 'GET'],
  url: '*',
  handler: async ({ raw: req }, { raw: res }) => {
    const result = await storage.prepareResponse(req.url, req);
    result.send(res);
  }
});

app.listen(3000)
  .then(() => {
    console.info('listening on http://localhost:3000');
  });
```

Using Koa (v2.x.x):
```js
const path = require('path');
const Koa = require('koa');
const { FileSystemStorage } = require('send-stream');

const app = new Koa<object>();
const storage = new FileSystemStorage(path.join(__dirname, 'assets'), { onDirectory: 'serve-index' });

app.use(async ctx => {
  let result = await storage.prepareResponse(ctx.request.path, ctx.req);
  ctx.response.status = result.statusCode;
  ctx.response.set(result.headers);
  ctx.body = result.stream;
});

app.listen(3000, () => {
  console.info('listening on http://localhost:3000');
});
```

Using Express (v4.x.x):
```js
const path = require("path");
const express = require("express");
const { FileSystemStorage } = require('send-stream');

const app = express();
const storage = new FileSystemStorage(path.join(__dirname, 'assets'), { onDirectory: 'serve-index' });

app.get('*', async (req, res, next) => {
  try {
    let result = await storage.prepareResponse(req.url, req);
    result.send(res);
  } catch (err) {
    next(err);
  }
});
app.listen(3000, () => {
  console.info('listening on http://localhost:3000');
});
```

## API

### `new FileSystemStorage(root, [options])`

Create a new `FileSystemStorage` which is a stream storage giving access to the files inside the given root folder.

The **`root`** parameter is a the absolute path from which the storage takes the files.

The **`options`** parameter let you add some addition options.

#### Options

##### mimeTypeLookup

In order to return the content type, the storage will try to guess the mime type thanks to the `mime-types` module
(see [mime-types module documentation](https://github.com/broofa/node-mime)).
This option override the mime-types lookup function which will be used for this purpose.

Example:

```js
new FileSystemStorage(
  directory,
  {
    mimeTypeLookup: filename => {
      if (filename.endsWith('.abc')) {
        return 'text/abc';
      }
      return undefined;
    }
  }
)
```

---
##### mimeTypeDefaultCharset

In order to return the content type, the storage will try to guess the mime type charset thanks to the `mime-types`
module (see [mime-types module documentation](https://github.com/broofa/node-mime)).
This option override the mime-types charset function which will be used for this purpose.

Example:

```js
new FileSystemStorage(
  directory,
  {
    mimeTypeDefaultCharset: mimeType => {
      if (mimeType === 'text/abc') {
        return 'UTF-8';
      }
      return undefined;
    }
  }
)
```

---
##### dynamicCompression

Enable dynamic compression of file content.
This can be a boolean or a list of encodings ordered by priority, `['br', 'gzip']` if `true` is used.
Activating this option will automatically compress content as brotli or gzip
if the content is detected as compressible and supported by the client.

Note that this is highly recommended to use this option only if you can not use pre-compressed options
like the 'contentEncodingMappings' option from FileSystemStorage.

Also when dynamic compression is active, `Content-Length` header will be removed
and range requests will be disabled as content length is unknown

Defaults to `false`

Example:

```js
new FileSystemStorage(
  directory,
  {
    dynamicCompression: true
  }
)
```

---
##### mimeTypeCompressible

Function used to determine if a type is compressible (for dynamic compression only)
`compressible` module will be used by default

Example:

```js
new FileSystemStorage(
  directory,
  {
    dynamicCompression: true,
    mimeTypeCompressible: mimeType => mimeType === 'text/plain'
  }
)
```

---
##### dynamicCompressionMinLength

Sets the minimum length of a response that will be dynamically compressed (only when the length is known)

Default to 20

Example:

```js
new FileSystemStorage(
  directory,
  {
    dynamicCompression: true,
    dynamicCompressionMinLength: 100
  }
)
```

---
##### defaultMimeType

Configures the default content type (without charset) that will be used if the content type is unknown.

`undefined` by default

Example:

```js
new FileSystemStorage(directory, { defaultMimeType: 'application/octet-stream' })
```

---
##### maxRanges

Configure how many ranges can be accessed in one request.

Defaults to `200`

Setting it to `1` will disable multipart/byteranges

Setting it to `0` or less will disable range requests

Example:

```js
new FileSystemStorage(directory, { maxRanges: 10 })
```

---
##### weakEtags

The storage will generate strong etags by default, when set to true the storage will generate weak etags instead.

`false` by default

Example:

```js
new FileSystemStorage(directory, { weakEtags: true })
```

---
##### contentEncodingMappings

Configure content encoding file mappings.

`undefined` by default

Example:

```js
new FileSystemStorage(
  directory,
  {
    contentEncodingMappings: [
      {
        matcher: /^(.+\.(?:html|js|css))$/,
        encodings: [
          { name: 'br', path: '$1.br' },
          { name: 'gzip', path: '$1.gz' }
        ]
      }
    ]
  }
)
```

---
##### ignorePattern

The storage will ignore files which have any parts of the path matching this pattern.

`false` will disable it

Defaults to `/^\./` (files/folders beginning with a dot)

Example:

```js
new FileSystemStorage(directory, { ignorePattern: /^myPrivateFolder$/ })
```

---
##### fsModule

Let you override the `fs` module that will be used to retrieve files.

Example:

```js
const memfs = require('memfs');
memfs.fs.writeFileSync('/hello.txt', 'world');
new FileSystemStorage(directory, { fsModule: memfs })
```
---
##### directory

Determine what should happen on directory requests (trailing slash)
 - `false` to return an error
 - `'list-files'` to list the files of directories
 - `'serve-index'` to serve the index.html file of directories

Default to false

Note that you can customize the html template used for `'list-files'` by overiding `getDirectoryListing` method.

Example:

```js
new FileSystemStorage(directory, { directory: 'list-files' })
```
---

### `storage.prepareResponse(path, req, [options])`

Create asynchronously a new `StreamResponse` for the given path relative to root ready to be sent to a server response.

The **`path`** parameter is a urlencoded path (urlencoded) or an array of path parts (should always start with '').

For example, `'/my%20directory/index.html'` is the equivalent of `['', 'my directory', 'index.html']`.

Query params will be ignored if present.

The **`req`** is the related request, it can be a `http.IncomingMessage` or a `http2.Http2ServerRequest` or a `http2.IncomingHttpHeaders`.

The **`options`** parameter let you add some addition options.

#### Options

##### cacheControl

Custom cache-control header value, overrides storage value (`public, max-age=0` by default)

`false` to remove header

Example:

```js
await storage.prepareResponse(req.url, req, { cacheControl: 'public, max-age=31536000' })
```

---
##### lastModified

Custom last-modified header value, overrides storage value (defaults to mtimeMs converted to UTC)

`false` to remove header

Example:

```js
await storage.prepareResponse(req.url, req, { lastModified: 'Wed, 21 Oct 2015 07:28:00 GMT' })
```

---
##### etag

Custom etag header value, overrides storage value (defaults to size + mtimeMs + content encoding)

`false` to remove header


History.md
2.1.0
==================

* Feature: Add dynamicCompressionMinLength option
* Fix: Add missing documentation on `dynamicCompression` and `mimeTypeCompressible`

2.0.0
==================

* Breaking change: rename `contentType` prepare option to `mimeType` and `contentTypeCharset` to `mimeTypeCharset`
* Breaking change: rename storage option `defaultContentType` to `defaultMimeType`
* Breaking change: replace `defaultCharsets` and `mimeModule` storage options with `mimeTypeLookup` and
`mimeTypeDefaultCharset` functions
* Breaking change: `contentEncoding` in storage information is now (and needs to be) undefined when identity is used
* Feature: `mime` package is replaced by `mime-types` for charset lookups
* Feature: Add `dynamicCompression` and `mimeTypeCompressible` storage options
* Feature: Add `cacheControl`, `contentDispositionType` and `contentDispositionType` in storage information
* Use fastify v3 in examples
* Use typescript project references
* Bump typescript from 3.9.x to 4.0.x

1.1.0
==================

* Feature: Add a way to define ETag and Last-Modified headers directly from the storage

1.0.0
==================

* Breaking change: Each error now have a specific class instead of a code
* Breaking change: Encoded path must start with / now
* Breaking change: contentType values are now splitted into contentType (without charset) and contentTypeCharset
* Not normalized paths are now a 404 instead of 301 like other errors (redirects was unsafe depending on use)
* Fix some content encoding edge cases
* Multiple refactoring
* Feature: add onDirectory option
* Feature: storages can now emit contentType/contentTypeCharset values directly

0.4.0
==================

* Enhance CI ans tests
* Replace `tslint` with `eslint`
* Enhance documentation and examples
* Change default ignore pattern from `/^\../` to `/^\./`
* Rename error code `forbidden_characters` to `forbidden_character`
* Rename error code `ignored_files` to `ignored_file`
* Restore `malformed_path` error
* Throw `ignored_file` error before `trailing_slash`
* Drop `responseClose` event
* Add textual regexp support in options

0.3.0
==================

* Fix multi-range streams on error
* Create multi-range streams when needed only
* Fix mongodb example
* Update dependencies

0.2.1
==================

* Remove node 10 support

0.2.0
==================

* Breaking change: storage now takes a range object (which can be undefined) instead of start/end parameter
* File size/modification date/name are now optional in storage result
* 100% code coverage
* All file opening errors are now 404
* A parameter have been added to allow http methods different than GET and HEAD

0.1.1
==================

* Add new linting rules
* Update dependencies

0.1.0
==================

* Remove 1 minute freshness condition
* Add badges in documentation

0.0.11
==================

* Fix some examples
* Enhance code coverage
* Fix minor issues

0.0.10
==================

* Remove max-age on error
* Enhance examples

0.0.9
==================

* Add some documentation on response and errors

0.0.8
==================

* Change fullReponse option to statusCode option
* Add clarification on query params in path
* Enhance mongodb/gridfs example

0.0.7
==================

* Change default charset from utf-8 to UTF-8
* Add public in cache control by default
* Remove options method handling

0.0.6
==================

* Add weakEtags option
* Add path parts documentation
* Add pre compressed example

0.0.5
==================

 * Fix content-encoding when pre-compressed files do not exist
 * Fix crypto warning

0.0.4
==================

 * Keep compatibility with disabled allowSyntheticDefaultImports

0.0.3
==================

 * Relax StreamResponse typings

0.0.2
==================

 * Add fullResponse option

0.0.1
==================

 * First release


src/storage.ts

import * as http from 'http';
import * as http2 from 'http2';
import type { Readable } from 'stream';
import { pipeline } from 'stream';
import * as zlib from 'zlib';

import contentDisposition from 'content-disposition';
import { lookup, charset } from 'mime-types';
import parseRange from 'range-parser';
import compressible from 'compressible';

import { StreamResponse } from './response';
import { EmptyStream, BufferStream, MultiStream } from './streams';
import type { ResponseHeaders, BufferOrStreamRange } from './utils';
import {
	millisecondsToUTCString,
	statsToEtag,
	getFreshStatus,
	isRangeFresh,
	contentRange,
	randomBytes,
	StreamRange,
	acceptEncodings,
} from './utils';
import type {
	StorageOptions,
	PrepareResponseOptions,
	StorageRequestHeaders,
	StorageInfo,
} from './storage-models';
import { StorageError } from './storage-models';

const DEFAULT_ALLOWED_METHODS = <const> ['GET', 'HEAD'];
const DEFAULT_MAX_RANGES = 200;

/**
 * send-stream storage base class
 */
export abstract class Storage<Reference, AttachedData> {
	readonly defaultMimeType: string | false;

	readonly maxRanges: number;

	readonly weakEtags: boolean;
	readonly mimeTypeLookup: NonNullable<StorageOptions['mimeTypeLookup']>;
	readonly mimeTypeDefaultCharset: NonNullable<StorageOptions['mimeTypeDefaultCharset']>;
	readonly dynamicCompression: {
		encodingPreferences: ReadonlyMap<string, { order: number }>;
		identityEncodingPreference: { order: number };
	} | false;
	readonly mimeTypeCompressible: NonNullable<StorageOptions['mimeTypeCompressible']>;
	readonly dynamicCompressionMinLength: number;

	/**
	 * Create storage
	 *
	 * @param opts - storage options
	 */
	constructor(opts: StorageOptions = { }) {
		this.mimeTypeLookup = opts.mimeTypeLookup ?? lookup;
		this.mimeTypeDefaultCharset = opts.mimeTypeDefaultCharset ?? charset;
		if (opts.dynamicCompression) {
			const encodingPreferences: Map<string, { order: number }>
				= opts.dynamicCompression === true
					? new Map([['br', { order: 0 }], ['gzip', { order: 1 }], ['identity', { order: 2 }]])
					: new Map(opts.dynamicCompression.map((name, order) => [name, { order }]));
			let identityEncodingPreference = encodingPreferences.get('identity');
			if (!identityEncodingPreference) {
				identityEncodingPreference = { order: encodingPreferences.size };
				encodingPreferences.set('identity', identityEncodingPreference);
			}
			this.dynamicCompression = { encodingPreferences, identityEncodingPreference };
		} else {
			this.dynamicCompression = false;
		}
		this.mimeTypeCompressible = opts.mimeTypeCompressible ?? compressible;
		this.dynamicCompressionMinLength = opts.dynamicCompressionMinLength ?? 20;
		this.defaultMimeType = opts.defaultMimeType ?? false;
		this.maxRanges = opts.maxRanges ?? DEFAULT_MAX_RANGES;
		this.weakEtags = opts.weakEtags === true;
	}

	/**
	 * Create last-mofified header value from storage information (uses mtimeMs)
	 *
	 * @param storageInfo - storage information
	 * @returns last-mofified header
	 */
	// eslint-disable-next-line class-methods-use-this
	createLastModified(storageInfo: StorageInfo<AttachedData>): string | false {
		const { lastModified } = storageInfo;
		if (lastModified) {
			return lastModified;
		}
		const { mtimeMs } = storageInfo;
		if (mtimeMs === undefined) {
			return false;
		}
		return millisecondsToUTCString(mtimeMs);
	}

	/**
	 * Create etag header value from storage information (uses mtimeMs, size and contentEncoding)
	 *
	 * @param storageInfo - storage information
	 * @returns etag header
	 */
	createEtag(storageInfo: StorageInfo<AttachedData>): string | false {
		const { etag } = storageInfo;
		if (etag) {
			return etag;
		}
		const { size, mtimeMs } = storageInfo;
		if (size === undefined || mtimeMs === undefined) {
			return false;
		}
		return statsToEtag(size, mtimeMs, storageInfo.contentEncoding, this.weakEtags);
	}

	/**
	 * Create cache-control header value from storage information (return always public, max-age=0 unless overriden)
	 *
	 * @param storageInfo - storage information
	 * @returns cache-control header
	 */
	// eslint-disable-next-line class-methods-use-this
	createCacheControl(storageInfo: StorageInfo<AttachedData>): string | false {
		const { cacheControl } = storageInfo;
		if (cacheControl) {
			return cacheControl;
		}
		return 'public, max-age=0';
	}

	/**
	 * Create mime type for content-type header value from storage information
	 *
	 * @param storageInfo - storage information (unused unless overriden)
	 * @returns mime type
	 */
	createMimeType(storageInfo: StorageInfo<AttachedData>): string | false {
		const { mimeType } = storageInfo;
		if (mimeType) {
			return mimeType;
		}
		const { fileName } = storageInfo;
		if (!fileName) {
			return this.defaultMimeType;
		}
		const type = this.mimeTypeLookup(fileName);
		if (!type) {
			return this.defaultMimeType;
		}
		return type;
	}

	/**
	 * Create charset that will be appended with mime type into content-type header
	 *
	 * @param storageInfo - storage information (unused unless overriden)
	 * @param mimeType - mime type
	 * @returns charset
	 */
	createMimeTypeCharset(storageInfo: StorageInfo<AttachedData>, mimeType: string): string | false {
		if (storageInfo.mimeTypeCharset) {
			return storageInfo.mimeTypeCharset;
		}
		return this.mimeTypeDefaultCharset(mimeType);
	}

	/**
	 * Create content-disposition header type from storage information (return always inline unless overriden)
	 *
	 * @param storageInfo - storage information (unused unless overriden)
	 * @returns content-disposition header type
	 */
	// eslint-disable-next-line class-methods-use-this
	createContentDispositionType(storageInfo: StorageInfo<AttachedData>): 'inline' | 'attachment' | undefined {
		const { contentDispositionType } = storageInfo;
		if (contentDispositionType) {
			return contentDispositionType;
		}
		return 'inline';
	}

	/**
	 * Create content-disposition header filename from storage information
	 * (return always the original filename unless overriden)
	 *
	 * @param storageInfo - storage information
	 * @returns content-disposition header filename
	 */
	// eslint-disable-next-line class-methods-use-this
	createContentDispositionFilename(storageInfo: StorageInfo<AttachedData>): string | undefined {
		const { contentDispositionFilename } = storageInfo;
		if (contentDispositionFilename) {
			return contentDispositionFilename;
		}
		return storageInfo.fileName;
	}

	/**
	 * Prepare to send file
	 *
	 * @param reference - file reference
	 * @param req - request headers or request objects
	 * @param [opts] - options
	 * @returns status, response headers and body to use
	 * @throws when method is incorrect or when storage can not create the storage stream
	 */
	async prepareResponse(
		reference: Reference,
		req: http.IncomingMessage | http2.Http2ServerRequest | http2.IncomingHttpHeaders,
		opts: PrepareResponseOptions = {},
	): Promise<StreamResponse<AttachedData>> {
		let method;
		let requestHeaders: StorageRequestHeaders;
		if (req instanceof http.IncomingMessage || req instanceof http2.Http2ServerRequest) {
			method = req.method;
			requestHeaders = req.headers;
		} else {
			method = req[':method'];
			requestHeaders = req;
		}
		if (!method) {
			throw new Error('cannot send, method is missing');
		}
		const isGetMethod = method === 'GET';
		const isHeadMethod = method === 'HEAD';
		const isGetOrHead = isGetMethod || isHeadMethod;
		const allowedMethods = opts.allowedMethods ?? DEFAULT_ALLOWED_METHODS;
		if (!allowedMethods.includes(method)) {
			return this.createMethodNotAllowedError(isHeadMethod, allowedMethods);
		}
		let earlyClose = false;
		let storageInfo;
		let contentLength;
		let dynamicContentEncoding;
		try {
			storageInfo = await this.open(
				reference,
				requestHeaders,
			);
		} catch (error: unknown) {
			return this.createStorageError(isHeadMethod, error);
		}
		try {
			const responseHeaders: ResponseHeaders = { };
			const mimeType = opts.mimeType ?? this.createMimeType(storageInfo);
			let mimeTypeCharset;
			if (mimeType) {
				storageInfo.mimeType = mimeType;
				mimeTypeCharset = opts.mimeTypeCharset ?? this.createMimeTypeCharset(storageInfo, mimeType);
				if (mimeTypeCharset) {
					storageInfo.mimeTypeCharset = mimeTypeCharset;
				}
			}
			const { dynamicCompression } = this;
			if (
				dynamicCompression
				&& !storageInfo.contentEncoding
				&& mimeType
				&& this.mimeTypeCompressible(mimeType)
				&& (storageInfo.size === undefined || storageInfo.size > this.dynamicCompressionMinLength)
			) {
				storageInfo.vary = 'Accept-Encoding';
				const [[preferedEncoding]] = acceptEncodings(
					requestHeaders['accept-encoding'],
					dynamicCompression.encodingPreferences,
					dynamicCompression.identityEncodingPreference,
				);
				if (preferedEncoding !== 'identity') {
					storageInfo.contentEncoding = preferedEncoding;
					dynamicContentEncoding = preferedEncoding;
				}
			}

			const lastModified = opts.lastModified ?? this.createLastModified(storageInfo);

			if (lastModified) {
				storageInfo.lastModified = lastModified;
			}

			const etag = opts.etag ?? this.createEtag(storageInfo);

			if (etag) {
				storageInfo.etag = etag;
			}

			const contentDispositionType = opts.contentDispositionType
				?? this.createContentDispositionType(storageInfo);
			let contentDispositionFilename;
			if (contentDispositionType) {
				storageInfo.contentDispositionType = contentDispositionType;
				const { contentDispositionFilename: optsContentDispositionFilename } = opts;
				contentDispositionFilename = optsContentDispositionFilename === undefined
					? this.createContentDispositionFilename(storageInfo)
					: optsContentDispositionFilename
						? optsContentDispositionFilename
						: undefined;
				storageInfo.contentDispositionFilename = contentDispositionFilename;
			}

			const cacheControl = opts.cacheControl ?? this.createCacheControl(storageInfo);
			if (cacheControl) {
				storageInfo.cacheControl = cacheControl;
				responseHeaders['Cache-Control'] = cacheControl;
			}

			if (storageInfo.vary) {
				responseHeaders['Vary'] = storageInfo.vary;
			}

			const fullResponse = opts.statusCode !== undefined;
			if (!fullResponse) {
				if (lastModified) {
					responseHeaders['Last-Modified'] = lastModified;
				}

				if (etag) {
					responseHeaders['ETag'] = etag;
				}

				const freshStatus = getFreshStatus(isGetOrHead, requestHeaders, etag, lastModified);
				switch (freshStatus) {
				case 304:
					earlyClose = true;
					return this.createNotModifiedResponse(responseHeaders, storageInfo);
				case 412:
					earlyClose = true;
					return this.createPreconditionFailedError(isHeadMethod, storageInfo);
				default:
					break;
				}
			}

			if (storageInfo.contentEncoding) {
				responseHeaders['Content-Encoding'] = storageInfo.contentEncoding;
			}

			let contentTypeHeader;
			if (mimeType) {
				if (mimeTypeCharset) {
					contentTypeHeader = `${ mimeType }; charset=${ mimeTypeCharset }`;
				} else {
					contentTypeHeader = mimeType;
				}
				responseHeaders['Content-Type'] = contentTypeHeader;
				responseHeaders['X-Content-Type-Options'] = 'nosniff';
			}

			if (contentDispositionType) {
				responseHeaders['Content-Disposition'] = contentDisposition(
					contentDispositionFilename,
					{ type: contentDispositionType },
				);
			}

			let statusCode = opts.statusCode ?? 200;
			const { size } = storageInfo;
			let rangeToUse: StreamRange | BufferOrStreamRange[] | undefined;
			if (size === undefined) {
				responseHeaders['Accept-Ranges'] = 'none';
				rangeToUse = undefined;
			} else {
				const { maxRanges } = this;
				if (maxRanges <= 0 || fullResponse || !isGetOrHead || dynamicContentEncoding) {
					responseHeaders['Accept-Ranges'] = 'none';
					rangeToUse = new StreamRange(0, size - 1);
					contentLength = size;
				} else {
					responseHeaders['Accept-Ranges'] = 'bytes';
					const { range: rangeHeader } = requestHeaders;
					if (
						!rangeHeader
						|| !isRangeFresh(requestHeaders, etag, lastModified)
					) {
						rangeToUse = new StreamRange(0, size - 1);
						contentLength = size;
					} else {
						const parsedRanges = parseRange(size, rangeHeader, { combine: true });
						if (parsedRanges === -1) {
							earlyClose = true;
							return this.createRangeNotSatisfiableError(isHeadMethod, size, storageInfo);
						}
						if (parsedRanges === -2
							|| parsedRanges.type !== 'bytes'
							|| parsedRanges.length > maxRanges
						) {
							rangeToUse = new StreamRange(0, size - 1);
							contentLength = size;
						} else {
							statusCode = 206;
							if (parsedRanges.length === 1) {
								const [singleRange] = parsedRanges;
								responseHeaders['Content-Range'] = contentRange('bytes', size, singleRange);
								rangeToUse = new StreamRange(singleRange.start, singleRange.end);
								contentLength = singleRange.end + 1 - singleRange.start;
							} else {
								const boundary = `----SendStreamBoundary${ (await randomBytes(24)).toString('hex') }`;
								responseHeaders['Content-Type'] = `multipart/byteranges; boundary=${ boundary }`;
								responseHeaders['X-Content-Type-Options'] = 'nosniff';
								rangeToUse = [];
								contentLength = 0;
								let first = true;
								for (const range of parsedRanges) {
									let header = `${ first ? '' : '\r\n' }--${ boundary }\r\n`;
									first = false;
									if (contentTypeHeader) {
										header += `content-type: ${ contentTypeHeader }\r\n`;
									}
									header += `content-range: ${ contentRange('bytes', size, range) }\r\n\r\n`;
									const headerBuffer = Buffer.from(header);
									rangeToUse.push(headerBuffer, new StreamRange(range.start, range.end));
									contentLength += headerBuffer.byteLength + range.end + 1 - range.start;
								}
								const footer = `\r\n--${ boundary }--`;
								const footerBuffer = Buffer.from(footer);
								rangeToUse.push(footerBuffer);
								contentLength += footerBuffer.byteLength;
							}
						}
					}
				}
				if (!dynamicContentEncoding) {
					responseHeaders['Content-Length'] = String(contentLength);
				}
			}

			let stream: Readable;
			if (isHeadMethod) {
				earlyClose = true;
				stream = new EmptyStream();
			} else if (rangeToUse === undefined) {
				stream = this.createReadableStream(storageInfo, undefined, true);
			} else if (rangeToUse instanceof StreamRange) {
				if (rangeToUse.end < rangeToUse.start) {
					earlyClose = true;
					stream = new EmptyStream();
				} else {
					stream = this.createReadableStream(storageInfo, rangeToUse, true);
				}
			} else {
				const si = storageInfo;
				stream = new MultiStream(
					rangeToUse,
					range => {
						if (range instanceof StreamRange) {
							return this.createReadableStream(
								si,
								range,
								false,
							);
						}
						return new BufferStream(range);
					},
					async () => this.close(si),
				);
			}
			if (dynamicContentEncoding) {
				stream = this.createCompressedStream(stream, dynamicContentEncoding, contentLength);
			}

			return this.createSuccessfulResponse(statusCode, responseHeaders, stream, storageInfo);
		} catch (err: unknown) {
			earlyClose = true;
			throw err;
		} finally {
			if (earlyClose) {
				await this.close(storageInfo);
			}
		}
	}

	// eslint-disable-next-line class-methods-use-this
	createCompressedStream(stream: Readable, contentEncoding: string, expectedSize?: number) {
		switch (contentEncoding) {
		case 'br':
			return pipeline(
				stream,
				zlib.createBrotliCompress({
					params: {
						[zlib.constants.BROTLI_PARAM_MODE]: zlib.constants.BROTLI_MODE_TEXT,
						[zlib.constants.BROTLI_PARAM_QUALITY]: 4,
						[zlib.constants.BROTLI_PARAM_SIZE_HINT]: expectedSize ?? 0,
					},
				}),
				err => {
					if (err) {
						console.error('Broti compress failed.', err);
					}
				},
			);
		case 'gzip':
			return pipeline(
				stream,
				zlib.createGzip({ level: 6 }),
				err => {
					if (err) {
						console.error('Gzip failed.', err);
					}
				},
			);
		default:
			throw new Error(`${
				contentEncoding
			} is not supported as dynamic compression encoding (you can override createCompressedStream to handle it)`);
		}
	}

	/**
	 * Create Method Not Allowed error response
	 *
	 * @param isHeadMethod - true if HEAD method is used
	 * @param allowedMethods - allowed methods for Allow header
	 * @returns Method Not Allowed response
	 */
	// eslint-disable-next-line class-methods-use-this
	createMethodNotAllowedError(isHeadMethod: boolean, allowedMethods: readonly string[]) {
		// Method Not Allowed
		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
		const statusMessageBuffer = Buffer.from(http.STATUS_CODES['405']!);
		return new StreamResponse<AttachedData>(
			405,
			{
				// eslint-disable-next-line @typescript-eslint/naming-convention
				'Content-Length': String(statusMessageBuffer.byteLength),
				// eslint-disable-next-line @typescript-eslint/naming-convention
				'Content-Type': 'text/plain; charset=UTF-8',
				// eslint-disable-next-line @typescript-eslint/naming-convention
				'X-Content-Type-Options': 'nosniff',
				// eslint-disable-next-line @typescript-eslint/naming-convention
				Allow: allowedMethods.join(', '),
			},
			isHeadMethod ? new EmptyStream() : new BufferStream(statusMessageBuffer),
		);
	}

	/**
	 * Create storage error response (Not Found response usually)
	 *
	 * @param isHeadMethod - true if HEAD method is used
	 * @param error - the error causing this response
	 * @returns the error response
	 */
	// eslint-disable-next-line class-methods-use-this
	createStorageError(isHeadMethod: boolean, error: unknown) {
		// Not Found
		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
		const statusMessageBuffer = Buffer.from(http.STATUS_CODES['404']!);
		return new StreamResponse<AttachedData>(
			404,
			{
				// eslint-disable-next-line @typescript-eslint/naming-convention
				'Content-Length': String(statusMessageBuffer.byteLength),
				// eslint-disable-next-line @typescript-eslint/naming-convention
				'Content-Type': 'text/plain; charset=UTF-8',
				// eslint-disable-next-line @typescript-eslint/naming-convention
				'X-Content-Type-Options': 'nosniff',
			},
			isHeadMethod ? new EmptyStream() : new BufferStream(statusMessageBuffer),
			undefined,
			error instanceof StorageError ? error : new StorageError('Unknown error', error),
		);
	}

	/**
	 * Create Not Modified response
	 *
	 * @param responseHeaders - response headers
	 * @param storageInfo - the current storage info
	 * @returns the Not Modified response
	 */
	// eslint-disable-next-line class-methods-use-this
	createNotModifiedResponse(
		responseHeaders: ResponseHeaders,
		storageInfo: StorageInfo<AttachedData>,
	) {
		// Not Modified
		return new StreamResponse(304, responseHeaders, new EmptyStream(), storageInfo);
	}

	/**
	 * Create the Precondition Failed error response
	 *
	 * @param isHeadMethod - true if HEAD method is used
	 * @param storageInfo - the current storage info
	 * @returns the Precondition Failed error response
	 */
	// eslint-disable-next-line class-methods-use-this
	createPreconditionFailedError(isHeadMethod: boolean, storageInfo: StorageInfo<AttachedData>) {
		// Precondition Failed
		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
		const statusMessageBuffer = Buffer.from(http.STATUS_CODES['412']!);
		return new StreamResponse(
			412,
			{
				// eslint-disable-next-line @typescript-eslint/naming-convention
				'Content-Type': 'text/plain; charset=UTF-8',
				// eslint-disable-next-line @typescript-eslint/naming-convention
				'X-Content-Type-Options': 'nosniff',
				// eslint-disable-next-line @typescript-eslint/naming-convention
				'Content-Length': String(statusMessageBuffer.byteLength),
			},
			isHeadMethod ? new EmptyStream() : new BufferStream(statusMessageBuffer),
			storageInfo,
		);
	}

	/**
	 * Create the Range Not Satisfiable error response
	 *
	 * @param isHeadMethod - true if HEAD method is used
	 * @param size - size of content for Content-Range header
	 * @param storageInfo - the current storage info
	 * @returns the Range Not Satisfiable error response
	 */
	// eslint-disable-next-line class-methods-use-this
	createRangeNotSatisfiableError(isHeadMethod: boolean, size: number, storageInfo: StorageInfo<AttachedData>) {
		// Range Not Satisfiable
		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
		const statusMessageBuffer = Buffer.from(http.STATUS_CODES['416']!);
		return new StreamResponse(
			416,
			{
				// eslint-disable-next-line @typescript-eslint/naming-convention
				'Content-Range': contentRange('bytes', size),
				// eslint-disable-next-line @typescript-eslint/naming-convention
				'Content-Type': 'text/plain; charset=UTF-8',
				// eslint-disable-next-line @typescript-eslint/naming-convention
				'X-Content-Type-Options': 'nosniff',
				// eslint-disable-next-line @typescript-eslint/naming-convention
				'Content-Length': String(statusMessageBuffer.byteLength),
			},
			isHeadMethod ? new EmptyStream() : new BufferStream(statusMessageBuffer),
			storageInfo,
		);
	}

	/**
	 * Create the successful OK (200) or Partial Content (206) response
	 * (the http code could also be the one set in parameters)
	 *
	 * @param statusCode - 200 or 206 or the statusCode set in parameters
	 * @param responseHeaders - the response headers
	 * @param stream - the content stream
	 * @param storageInfo - the current storage info
	 * @returns the successful response
	 */
	// eslint-disable-next-line class-methods-use-this
	createSuccessfulResponse(
		statusCode: number,
		responseHeaders: ResponseHeaders,
		stream: Readable,
		storageInfo: StorageInfo<AttachedData>,
	) {
		// Ok | Partial Content
		return new StreamResponse(statusCode, responseHeaders, stream, storageInfo);
	}

	/**
	 * Open file and retrieve storage information (filename, modification date, size, ...)
	 *
	 * @param reference - file reference
	 * @param requestHeaders - request headers
	 * @returns StorageInfo object
	 */
	abstract open(
		reference: Reference,
		requestHeaders: StorageRequestHeaders,
	): Promise<StorageInfo<AttachedData>>;

	/**
	 * Create readable stream from storage information
	 *
	 * @param storageInfo - storage information
	 * @param range - range to use or undefined if size is unknown
	 * @param autoClose - true if stream should close itself
	 * @returns readable stream
	 */
	abstract createReadableStream(
		storageInfo: StorageInfo<AttachedData>,
		range: StreamRange | undefined,
		autoClose: boolean
	): Readable;

	/**
	 * Close storage information (if needed)
	 *
	 * @param storageInfo - storage information
	 */
	abstract close(storageInfo: StorageInfo<AttachedData>): Promise<void>;
}
